File: app/src-tauri/src/domain/message.rs
// Message domain types — re-exported from the orqa-engine crate.
//
// Message, MessageRole, ContentType, StreamStatus, MessageId, and SearchResult
// represent chat messages within a session. These types flow from the database
// through the Tauri IPC boundary to the frontend.

pub use orqa_engine::types::message::*;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn message_role_serializes_snake_case() {
        assert_eq!(
            serde_json::to_value(MessageRole::User)
                .expect("serialization should succeed")
                .as_str(),
            Some("user")
        );
        assert_eq!(
            serde_json::to_value(MessageRole::Assistant)
                .expect("serialization should succeed")
                .as_str(),
            Some("assistant")
        );
        assert_eq!(
            serde_json::to_value(MessageRole::System)
                .expect("serialization should succeed")
                .as_str(),
            Some("system")
        );
    }

    #[test]
    fn content_type_serializes_snake_case() {
        assert_eq!(
            serde_json::to_value(ContentType::Text)
                .expect("serialization should succeed")
                .as_str(),
            Some("text")
        );
        assert_eq!(
            serde_json::to_value(ContentType::ToolUse)
                .expect("serialization should succeed")
                .as_str(),
            Some("tool_use")
        );
        assert_eq!(
            serde_json::to_value(ContentType::ToolResult)
                .expect("serialization should succeed")
                .as_str(),
            Some("tool_result")
        );
        assert_eq!(
            serde_json::to_value(ContentType::Thinking)
                .expect("serialization should succeed")
                .as_str(),
            Some("thinking")
        );
        assert_eq!(
            serde_json::to_value(ContentType::Image)
                .expect("serialization should succeed")
                .as_str(),
            Some("image")
        );
    }

    #[test]
    fn stream_status_serializes_snake_case() {
        assert_eq!(
            serde_json::to_value(StreamStatus::Pending)
                .expect("serialization should succeed")
                .as_str(),
            Some("pending")
        );
        assert_eq!(
            serde_json::to_value(StreamStatus::Complete)
                .expect("serialization should succeed")
                .as_str(),
            Some("complete")
        );
        assert_eq!(
            serde_json::to_value(StreamStatus::Error)
                .expect("serialization should succeed")
                .as_str(),
            Some("error")
        );
    }

    #[test]
    fn message_roundtrip() {
        let msg = Message {
            id: 1,
            session_id: 1,
            role: MessageRole::User,
            content_type: ContentType::Text,
            content: Some("Hello, how are you?".to_string()),
            tool_call_id: None,
            tool_name: None,
            tool_input: None,
            tool_is_error: false,
            turn_index: 0,
            block_index: 0,
            stream_status: StreamStatus::Complete,
            input_tokens: Some(10),
            output_tokens: None,
            created_at: "2026-03-03T00:00:00Z".to_string(),
        };

        let json = serde_json::to_string(&msg).expect("serialization should succeed");
        let deserialized: Message =
            serde_json::from_str(&json).expect("deserialization should succeed");

        assert_eq!(deserialized.id, msg.id);
        assert_eq!(deserialized.role, MessageRole::User);
        assert_eq!(deserialized.content_type, ContentType::Text);
        assert_eq!(deserialized.stream_status, StreamStatus::Complete);
    }

    #[test]
    fn search_result_serialization() {
        let result = SearchResult {
            message_id: 5,
            session_id: 1,
            session_title: Some("Debug session".to_string()),
            content: "Fix the parsing bug".to_string(),
            highlighted: "Fix the <mark>parsing</mark> bug".to_string(),
            rank: 0.95,
        };

        let json = serde_json::to_value(&result).expect("serialization should succeed");
        assert_eq!(json["message_id"], 5);
        assert_eq!(json["rank"], 0.95);
        assert!(json["highlighted"]
            .as_str()
            .expect("highlighted should be a string")
            .contains("<mark>"));
    }
}

File: libs/sdk/src/testing/builders/message.ts
/**
 * Test builder for conversation Message objects.
 *
 * Produces valid messages matching the orqa-studio Message shape.
 * The @orqastudio/types dependency will be added later — for now,
 * the interface is defined locally.
 */

/** Minimal Message shape for test building. */
export interface Message {
	id: number;
	session_id: number;
	role: "user" | "assistant" | "system";
	content_type: string;
	content: string;
	tool_call_id: string | null;
	tool_name: string | null;
	tool_input: string | null;
	tool_is_error: boolean;
	turn_index: number;
	block_index: number;
	stream_status: string;
	input_tokens: number | null;
	output_tokens: number | null;
	created_at: string;
}

const defaults: Message = {
	id: 1,
	session_id: 1,
	role: "user",
	content_type: "text",
	content: "Hello",
	tool_call_id: null,
	tool_name: null,
	tool_input: null,
	tool_is_error: false,
	turn_index: 0,
	block_index: 0,
	stream_status: "complete",
	input_tokens: null,
	output_tokens: null,
	created_at: "2026-01-01T00:00:00Z",
};

/**
 * Create a test Message with sensible defaults.
 *
 * ```ts
 * const msg = createMessage({ role: "assistant", content: "Hi!" });
 * ```
 * @param overrides
 */
export function createMessage(overrides: Partial<Message> = {}): Message {
	return { ...defaults, ...overrides };
}
